src/net/rpc.rs

```diff
@@ -1267,14 +1269,28 @@ impl ConversationHttp {
                     cause: None,
                 },
             ),
-            Ok(Some(Err(e))) => HttpResponseType::CallReadOnlyFunction(
-                response_metadata,
-                CallReadOnlyResponse {
-                    okay: false,
-                    result: None,
-                    cause: Some(e.to_string()),
-                },
-            ),
+            Ok(Some(Err(e))) => match e {
+                Unchecked(CheckErrors::CostBalanceExceeded(actual_cost, _))
```

Comment:
I just want to clarify the "path" you're expecting `execute_contract` to follow to get to this branch of the match statement (where we get the `CostBalanceExceeded` error). 

To me, it seems you want `execute_contract` to fail at its call to `runtime_cost`, which will check that the execution costs do not exceed those set by the limits the environment has set. However, the limit is equal to `options.read_only_call_limit` (as set in line 1250 of this file), and the value for this struct (type `ExecutionCost`) ultimately comes from `ConnectionOptions`, which is specified in the `Config` passed to initialize a new node. So, it seems like there is not a guarantee that `write_count` or `write_length` would be 0, meaning the call to `runtime_cost` in `execute_contract` may be okay even if there is a write.

Reply:
I also had a question about this.  I could actually see this going either way:

* This API endpoint explicitly forbids any writes, and causes a `CostBalanceExceeded` error the moment the code attempts to wrote.
* This API endpoint explicitly permits writes, but never materializes them.  This could be used to allow clients to simulate what would happen if they called a function that performed a write, even though the write never commits.

I can see value in both approaches.

Reply:
as described in #1641 

> Make it so that the RPC endpoint used for calling read-only functions can be used to call any functions. Make it so that any mutations are discarded

this makes it possible to call any function, and get a result if no write occurred.  Furthermore, a read-only function that makes a `contrac-call?` is flagged as non read only, even if it is calling other read-only functions and the `call-read` endpoint will fail.  So this only allow calling functions that are so basic it is not so useful in practice, and it prevents composing contracts.  You have to use `define-public` instead, but this is not callable via the `call-read` endpoint.

So this extends the `call-read` endpoint to try the function, and return a result if no writes occurred, or a failure if any writes did occur.

So, @jcnelson, this is definitely the former.

@pavitthrap I'm not sure I understand how the node configuration would come into the picture.  Either `write_count` is 0 after the execution and that's a read-only call, or it will fail because it was not read-only.

Reply:
I think @pavitthrap's concern (correct me if I'm wrong) is that `options.read_only_call_limit` may have a non-zero values, since this `ExecutionCost` instance is instantiated from the node's config file, and you're allowed to specify non-zero write sizes.  I think the intention for even permitting this at all is that one day, there would be an interface that lets you call read/write contract-calls such that the writes never materialize (or even hit disk).

I think the solution here would be to constract an `ExecutionCost` from `options.read_only_call_limit` and explicitly set the `write_length` and `write_count` fields to 0.

Reply:
oh, @pavitthrap comment makes sense now.  I wasn't expecting these values to be other that 0, but to be safe, I'll set them to 0.

Reply:
addressed in 57b29c8eece8f3d9ef89bb5fcb2ad51847dfbb7f